lib/src/git.rs

```diff
@@ -1324,66 +1331,166 @@ fn push_refs(
     })?;
     let mut remaining_remote_refs_expected_locations =
         qualified_remote_refs_expected_locations.clone();
-    let mut push_options = git2::PushOptions::new();
-    let mut proxy_options = git2::ProxyOptions::new();
-    proxy_options.auto();
-    push_options.proxy_options(proxy_options);
-    let mut callbacks = callbacks.into_git();
-    callbacks.push_negotiation(|updates| {
-        for update in updates {
-            let dst_refname = update
-                .dst_refname()
-                .expect("Expect reference name to be valid UTF-8");
-            let expected_location = (*qualified_remote_refs_expected_locations
-                .get(dst_refname)
-                .expect("Push is trying to move a ref it wasn't asked to move"))
-            .clone();
-            let expected_location_oid = expected_location
-                .map(|id| git2::Oid::from_bytes(id.as_bytes()).unwrap())
-                .unwrap_or(git2::Oid::zero());
-            // `update.src()` is the current actual position of the branch
-            // `dst_refname` on the remote. `update.dst()` is the position
-            // we are trying to push the branch to (AKA our local branch
-            // location). 0000000000 is a valid value for either, indicating
-            // branch creation or deletion.
-
-            // Short-term TODO: Replace this with actual rules of when to permit the push
-            tracing::info!(
-                "Forcing {dst_refname} to {dst}. It was at {src} at the server. We expected it to \
-                 be at {expected_location_oid}.",
-                src = update.src(),
-                dst = update.dst()
-            );
-        }
-        Ok(())
-    });
-    callbacks.push_update_reference(|refname, status| {
-        // The status is Some if the ref update was rejected
-        if status.is_none() {
-            remaining_remote_refs_expected_locations.remove(refname);
-        }
-        Ok(())
-    });
-    push_options.remote_callbacks(callbacks);
-    remote
-        .push(refspecs, Some(&mut push_options))
-        .map_err(|err| match (err.class(), err.code()) {
-            (git2::ErrorClass::Reference, git2::ErrorCode::NotFastForward) => {
-                GitPushError::NotFastForward
+    let mut failed_push_negotiations = vec![];
+    let push_result = {
+        let failed_push_negotiations_ref = &mut failed_push_negotiations;
```

Comment:
nit: no need to create `&mut ` reference

Reply:
Good point! I'm quite confused by what the compiler does and does not do automatically; I thought this fixed an error but I guess something else I did actually fixed it.